File: 2015/java/3/part1/Main.java
package Part1;

import java.io.File;
import java.util.Scanner;
import java.util.Map;
import java.util.HashMap;

import Position.Position;

public class Main {

    public static void main(String[] args) {
        if (args.length < 1) {
            System.out.println("Usage: java Main <input_file>");
            System.exit(1);
        }

        String input = args[0];

        try {
            File file = new File(input);
            Scanner scanner = new Scanner(file);

            while (scanner.hasNextLine()) {
                String line = scanner.nextLine();

                if (line.length() <= 0) {
                    continue;
                }

                Position pos = new Position(0, 0);

                Map<Position, Integer> map = new HashMap<>();
                map.put(pos, 1);

                for (int i = 0; i < line.length(); i++) {
                    Position newPos = new Position(pos.x, pos.y);
                    char direction = line.charAt(i);
                    newPos.move(direction);
                    newPos.reHash();
                    map.put(newPos, map.getOrDefault(newPos, 0) + 1);
                    pos = newPos;
                }

                int count = map.size();
                System.out.println("Part 1: " + count);
            }

            scanner.close();

        } catch (Exception e) {
            System.out.println("Error: " + e.getMessage());
        }
    }
}

File: 2015/rust/3/src/part1/main.rs
trait Movement {
    fn move_pos(&mut self, direction: char);
}

struct Position {
    x: i32,
    y: i32,
}

impl Movement for Position {
    fn move_pos(&mut self, direction: char) {
        match direction {
            '^' => self.y += 1,
            'v' => self.y -= 1,
            '<' => self.x -= 1,
            '>' => self.x += 1,
            _ => panic!("Invalid direction"),
        }
    }
}

fn main() {
    // get command line arguments
    let args: Vec<String> = std::env::args().collect();
    // check if there are enough arguments
    if args.len() < 2 {
        eprintln!("Usage: {} <input_file>", args[0]);
        std::process::exit(1);
    }

    // read the input file
    let input = std::fs::read_to_string(&args[1]).expect("Failed to read input file");

    // split the input into lines
    let lines: Vec<&str> = input.lines().collect();

    let mut pos = Position { x: 0, y: 0 };

    // create hashmap to store the number of houses visited
    let mut houses = std::collections::HashMap::new();
    // insert the starting position
    houses.insert((pos.x, pos.y), 1);

    for c in lines[0].chars() {
        pos.move_pos(c);
        let count = houses.entry((pos.x, pos.y)).or_insert(1);
        *count += 1;
    }

    let mut count = 0;
    for _house in houses.iter() {
        count += 1;
    }
    println!("Part 1: {}", count);
}
